### DISCUSSION TITLE
About the new RateLimiting annotation

### DISCUSSION THREAD

**Discussion by fabiofs-br** [MarcGiffing/bucket4j-spring-boot-starter#265] [Q&A] [answered]
Hi everybody. I'm pretty interested in the new RateLimiting annotation (#250). Seems perfect in my current project, where I with to limit few specific methods with a separated bucket logic from the rest.
I added version 0.12.0 dependency and tried to configure it as described in the documentation, but didn't manage to make it work. Does require any additional configuration?
I have some doubts about it functioning too. Like, does it override other configured filters?

**Comment by MarcGiffing** [comment]
Hey, the RateLimiting annotation is independent from the filter configuration.

Annotation:
`bucket4j.methods[0]`
Filter:
`bucket4j.filters[0]`

There is still a bug as it is only works in a web application:
https://github.com/MarcGiffing/bucket4j-spring-boot-starter/issues/260

Can you enable debug in your property and search for the Bucket4jAopConfig class. Is it enabled?
```
# property
debug=true
```

**Comment by fabiofs-br** [comment]
Yes. It's actually enabled.
```
Bucket4jAopConfig matched:
      - @ConditionalOnClass found required class 'org.aspectj.lang.annotation.Aspect' (OnClassCondition)
      - @ConditionalOnProperty (bucket4j.enabled) matched (OnPropertyCondition)
      - @ConditionalOnBean (types: com.giffing.bucket4j.spring.boot.starter.config.cache.SyncCacheResolver; SearchStrategy: all) found bean 'bucket4jCacheResolver' (OnBeanCondition)
``` 

Its parsing the configured expressions too, but no rate limit is applied to response of the controller.
```
parse-boolean-expression;result:true;expression:true;root:public org.springframework.http.ResponseEntity com.example.controller.Controller.method(java.lang.Long);params:{parameter=1}
parse-string-expression;result:1;expression:1;root:public org.springframework.http.ResponseEntity com.example.controller.Controller.method(java.lang.Long);params:{parameter=1}
``` 

I disabled the filters configuration while testing to prevent any conflict too.

Edit: After some more testing, I discover which the rate is actually been limited. Just not showing "X-Rate-Limit-Retry-After-Seconds" header in the response. Instead, it launches a RateLimitException which can be conveniently handled.

**Reply by MarcGiffing** [reply, marked as answer]
Yes, the headers are only added when the filters are used.
[Exception Handler in Caffeine Example](https://github.com/MarcGiffing/bucket4j-spring-boot-starter/blob/master/examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/RateLimitExceptionHandler.java)
